prometheus/promhttp/instrument_server.go

```diff
@@ -0,0 +1,433 @@
+// Copyright 2016 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Copyright (c) 2013, The Prometheus Authors
+// All rights reserved.
+//
+// Use of this source code is governed by a BSD-style license that can be found
+// in the LICENSE file.
+
+package promhttp
+
+import (
+	"bufio"
+	"io"
+	"net"
+	"net/http"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+	dto "github.com/prometheus/client_model/go"
+)
+
+// InstrumentHandlerInFlight accepts a Gauge and an http.Handler, returning a
+// new http.Handler that wraps the supplied http.Handler. The provided Gauge
+// must be registered in a registry in order to be used.
+func InstrumentHandlerInFlight(g prometheus.Gauge, next http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		g.Inc()
+		next.ServeHTTP(w, r)
+		g.Dec()
+	})
+}
+
+// InstrumentHandlerDuration accepts an ObserverVec interface and an
+// http.Handler, returning a new http.Handler that wraps the supplied
+// http.Handler. The provided ObserverVec must be registered in a registry in
+// order to be used. If the wrapped http.Handler has not set a status code,
+// i.e. the value is currently 0, the supplied ObserverVec will report a 200.
+// The instance labels "code" and "method" are supported on the provided
+// ObserverVec.  Note: Partitioning histograms is expensive.
+func InstrumentHandlerDuration(obs prometheus.ObserverVec, next http.Handler) http.HandlerFunc {
```

Comment:
In the same spirit as above. I would not invest effort into describing the signature of the function or mentioning that the ObserverVec must be registered to do anything. Instead, I would focus on the not very obvious way of treating the labels. Suggestion (not very polished):

"""
InstrumentHandlerDuration is a middleware that wraps the provided http.Handler to observe the request duration with the provided ObserverVec. The ObserverVec must have zero, one, or two labels, and the only allowed label names are "code" and "method". Otherwise, the function panics. The Observe method of the Observers in the ObserverVec is called with the request duration in seconds. Partitioning happens by HTTP status code and/or HTTP method if the respective label names are present in the ObserverVec. For unpartitioned observations, use an ObserverVec with zero labels. Note that partitioning of Histograms is expensive and should be used judiciously.

If the wrapped Handler does not set a status code, a status code of 200 is assumed.
"""

(Note the blank line as godoc needs it for a new paragraph.)